docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,356 @@
+  - We ensure that within *each* artifact dependency slice the aliases are unique. 
+
+## Referencing dependencies
+
+### Docker builder
+
+The `docker` builder will use the `alias` of an `ArtifactDependency` as a build argument key.
+ 
+ ```yaml
+ build:
+  artifacts:
+    - image: simple-go-app
+    - image: leeroy-app
+      requires:
+        - image: simple-go-app
+          alias: BASE
+ ```
+
+Here `BASE=gcr.io/X/simple-go-app:<tag>@sha:<sha>` is passed as a `--build-arg` (or a buildKit parameter) when building `leeroy-app`.
+
+### Custom builder
+
+The `custom` builder will be supplied each `ArtifactDependency`'s `alias` and image reference as environment variables keyed on `alias`. So they can be easily referenced in user-defined build definitions.
+
+### Buildpacks builder
+
+Buildpacks supports overriding the run-image and the builder-image in its current schema. We extend this to allow specifying `ArtifactDependency` aliases as the value for the `runImage` and `builder` fields.
+
+```yaml
+build:
+  artifacts:
+  - image: skaffold-buildpacks
+    buildpacks:
+      builder: "{{ .BUILDER_IMAGE }}"
+      runImage: "{{ .RUN_IMAGE }}"
+    requires:
+      - image: builder-image
+        alias: BUILDER_IMAGE
+      - image: run-image
+        alias: RUN_IMAGE
+```
+
+If there are any additional images in the `required` section it only enforces that they get built prior to the current image. However, the buildpacks builder cannot really reference them in any other way.
+
+### Jib builder
+
+The Jib builder supports [changing the base image](https://cloud.google.com/java/getting-started/jib#base-image). We allow substituting the `required` image aliases with the image references in the specific builder `args` section.
+This will allow overriding the `jib.from.image` property that sets the base image.
+
+For Maven:
+
+```yaml
+build:
+  artifacts:
+  - image: test-jib-maven
+    jib:
+      type: maven
+      args: 
+      - -DbaseImage=docker://{{ .BASE_IMAGE }}
+```
+
+where `baseImage` is added as a property to the `pom.xml`
+
+```xml
+<properties>
+  <baseImage>image</baseImage>
+</properties>
+...
+<configuration>
+  <from>
+    <image>${baseImage}</image>
+  </from>
+</configuration>
+```
+
+Similarly, for Gradle:
+
+```yaml
+build:
+  artifacts:
+  - image: test-jib-gradle
+    jib:
+      type: gradle
+      args: 
+      - -Djib.from.image=registry://{{ .BASE_IMAGE }}
+```
+
+where `jib.from.image` is added to the Gradle configuration.
+
+### Bazel builder
+
+The bazel builder doesn't support referencing images directly. Also, it natively supports setting up nested builds. We allow defining `required` artifacts even though they can't be referenced by the builder. We do this to give the user a way of ordering these builds; with the future work around [Skaffold Hooks](https://github.com/GoogleContainerTools/skaffold/issues/1441) there might be some usecases where pre and post build scripts might want a certain ordering of builds. 
```

Comment:
I'm on the fence about allowing `required` to provide implicit ordering.

Reply:
PTAL at comment [above](https://github.com/GoogleContainerTools/skaffold/pull/4794/files#r494065590)